File: leet/0055_jump_game/0055_jump_game.py
# LeetCode 55. Jump Game
# https://leetcode.com/problems/jump-game/
# O(n) time, O(1) space
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        i, maxJump = 0, 0
        while i <= maxJump:
            maxJump = max(maxJump, i + nums[i])
            if i >= len(nums) - 1:
                return True
            i += 1
        return False

File: leet/0055_jump_game/0055_jump_game.ts
// LeetCode 55. Jump Game
// https://leetcode.com/problems/jump-game/
// O(n) time, O(1) space
function canJump(nums: number[]): boolean {
    let maxJump = nums[0];
    for (let i = 1; i <= maxJump && i < nums.length; i++) {
        maxJump = Math.max(maxJump, nums[i] + i);
    }
    return maxJump >= nums.length - 1;
};

